api/handlers/listhandler.py

```diff
@@ -576,3 +584,132 @@ def packfile_end(self, cont_name, **kwargs):
+            self.response.headers['Content-Length'] = str(fileinfo['size']) # must be set after setting app_iter
+            if self.is_true('view'):
+                self.response.headers['Content-Type'] = str(fileinfo.get('mimetype', 'application/octet-stream'))
+            else:
+                self.response.headers['Content-Type'] = 'application/octet-stream'
+                self.response.headers['Content-Disposition'] = 'attachment; filename=' + str(filename)
+
+    def _prepare_batch(self, fileinfo):
+        ## duplicated code from download.py
+        ## we need a way to avoid this
+        targets = []
+        total_size = total_cnt = 0
+        data_path = config.get_item('persistent', 'data_path')
+        for f in fileinfo:
+            filepath = os.path.join(data_path, util.path_from_hash(f['hash']))
+            log.error(filepath)
+            if os.path.exists(filepath): # silently skip missing files
+                targets.append((filepath, 'analyses/' + f['name'], f['size']))
+                total_size += f['size']
+                total_cnt += 1
+        return targets, total_size, total_cnt
+
+    def _send_batch(self, ticket):
+        self.response.app_iter = download.archivestream(ticket)
+        self.response.headers['Content-Type'] = 'application/octet-stream'
+        self.response.headers['Content-Disposition'] = 'attachment; filename=' + str(ticket['filename'])
+
+    def delete_note(self, cont_name, list_name, **kwargs):
+        _id = kwargs.pop('cid')
+        analysis_id = kwargs.pop('_id')
+        container, permchecker, storage, _, _, _ = self._initialize_request(cont_name, list_name, _id)
+        note_id = kwargs.get('note_id')
+        permchecker(noop)('DELETE', _id=_id)
+        result = storage.delete_note(_id=_id, analysis_id=analysis_id, note_id=note_id)
+        if result.modified_count == 1:
+            return {'modified':result.modified_count}
+        else:
+            self.abort(404, 'Element not removed from list {} of container {} {}'.format(storage.list_name, storage.cont_name, _id))
+
+    def add_note(self, cont_name, list_name, **kwargs):
+        _id = kwargs.pop('cid')
+        analysis_id = kwargs.get('_id')
+        container, permchecker, storage, mongo_validator, input_validator, keycheck = self._initialize_request(cont_name, list_name, _id)
+        payload = self.request.json_body
+        input_validator(payload, 'POST')
+        payload['_id'] = str(bson.objectid.ObjectId())
+        payload['user'] = payload.get('user', self.uid)
+        payload['created'] = datetime.datetime.utcnow()
+        permchecker(noop)('POST', _id=_id)
+        result = storage.add_note(_id=_id, analysis_id=analysis_id, payload=payload)
+        if result.modified_count == 1:
+            return {'modified':result.modified_count}
+        else:
+            self.abort(404, 'Element not added in list {} of container {} {}'.format(storage.list_name, storage.cont_name, _id))
```

Comment:
Hmm, that makes more sense. I am not yet sold on the sub-document part, but it might be that I don't fully understand the feature. This makes more sense given how it's currently designed.